scale-out-relay/src/main/scala/com/normation/plugins/scaleoutrelay/ScaleOutRelayService.scala

```diff
@@ -0,0 +1,192 @@
+
+
+object ScaleOutRelayService {
+  val SYSTEM_GROUPS = "SystemGroups"
+  val DISTRIBUTE_POLICY = "distributePolicy"
+  val COMMON = "common"
+
+  private val nodeInfosService = RudderConfig.nodeInfoService
+  private val woLDAPNodeGroupRepository = RudderConfig.woNodeGroupRepository
+  private val woLDAPNodeRepository = RudderConfig.woNodeRepository
+  private val woDirectiveRepository = RudderConfig.woDirectiveRepository
+  private val woRuleRepository = RudderConfig.woRuleRepository
+  private val uuidGen = RudderConfig.stringUuidGenerator
+  private val policyServerManagementService = RudderConfig.policyServerManagementService
+  private val removeNodeService = RudderConfig.removeNodeService
+
+  def promoteNodeToRelay(uuid: NodeId, actor: EventActor, reason:Option[String]): ZIO[Any, RudderError, NodeInfo] = {
+    val modId = ModificationId(uuidGen.newUuid)
+    for {
+      nodeInfos   <- getNodeToPromote(uuid)
+      updatedNode =  createNodeToPolicyServerType(nodeInfos)
+      _           <- removeNodeService.removeNode(nodeInfos.node.id, modId, actor).toIO
+      newRelay <- createRelayFromNode(updatedNode, modId, actor, reason)
+    } yield {
+      newRelay
+    }
+  }
+
+  private[scaleoutrelay] def createRelayFromNode(nodeInf: NodeInfo, modId: ModificationId, actor: EventActor, reason:Option[String])= {
+    for {
+      commonDirective <- createCommonDirective(nodeInf).toIO
+
+      directDistribPolicy = createDirectiveDistributePolicy(nodeInf.node.id)
+      ruleTarget = createPolicyServer(nodeInf.node.id)
+      nodeGroup = createNodeGroup(nodeInf.node.id)
+      ruleDistribPolicy = createRuleDistributePolicy(nodeInf.node.id)
+      ruleSetup = createRuleSetup(nodeInf.node.id)
+
+      categoryId = NodeGroupCategoryId(SYSTEM_GROUPS)
+      activeTechniqueId = ActiveTechniqueId(DISTRIBUTE_POLICY)
+      activeTechniqueIdCommon = ActiveTechniqueId(COMMON)
+
+      _ <- woLDAPNodeRepository.createNode(nodeInf.node, modId, actor, reason)
+      _ <- woLDAPNodeGroupRepository.createPolicyServerTarget(ruleTarget, modId, actor, reason)
+      _ <- woLDAPNodeGroupRepository.create(nodeGroup, categoryId, modId, actor, reason)
+      _ <- woDirectiveRepository.saveSystemDirective(activeTechniqueId, directDistribPolicy, modId, actor, reason)
+      _ <- woDirectiveRepository.saveSystemDirective(activeTechniqueIdCommon, commonDirective, modId, actor, reason)
+      _ <- woRuleRepository.create(ruleSetup, modId, actor, reason)
+      _ <- woRuleRepository.create(ruleDistribPolicy, modId, actor, reason)
+    } yield {
+      nodeInf
+    }
+  }
+
+  private[scaleoutrelay] def getNodeToPromote(uuid: NodeId): IOResult[NodeInfo] = {
+    nodeInfosService.getNodeInfo(uuid) match {
+      case Full(nodeOpt) => nodeOpt.notOptional(s"Node with UUID ${uuid.value} is missing and can not be upgraded to relay")
+      case eb: EmptyBox => eb.toIO
+    }
+  }
+
+  private[scaleoutrelay] def createNodeToPolicyServerType(nodeInf: NodeInfo) = {
```

Comment:
In place of that, you can use quicklens (https://github.com/softwaremill/quicklens), it allows to modify values in immutable structure (it doesn't modify original, just return a modified copy):
```
nodeInf
  .modify(_.node.isSystem).setTo(true)
  .modify(_.node.isPolicyServer).setTo(true)
```
Moreover, I'm not sure why you set `id` and `name`: the value are alrady correct, no?